lib/api/northbound-api.js

```diff
@@ -0,0 +1,211 @@
+                });
+        } else{
+            throw new Errors.BadRequestError("Path is not readable");
+        }
+    }
+
+    function getImageQuery(req){
+        var query = {};
+
+        query.id = req.query.id;
+        query.name = req.query.name;
+        query.version = req.query.version;
+        query.iso = req.query.iso;
+        query.status = req.query.status;
+
+        return _(query).omit(_.isUndefined).omit(_.isNull).value();
+    }
+
+    NorthboundApi.prototype.getRouter = function getRouter(){
+        isoDirPath = path.join(process.cwd(),
+            configuration.get('isoDir', "./static/iso")
+        );
+
+        microkernelDirPath = path.join(process.cwd(),
+            configuration.get('microkernelDir', "./static/common")
+        );
+
+        inventory.initialize();
+
+        router.get('/images', rest(function(req) {
+            return inventory.findImagesByQuery(getImageQuery(req));
+        }));
+
+        router.put('/images', rest(function(req, res) {
+            req.query = req.query || {};
+            req.body = req.body || {};
+
+            var name = req.query.name || req.body.name;
+            var version = req.query.version || req.body.version;
+
+            var isostore = req.query.isostore || req.body.isostore;
+            var isolocal = req.query.isolocal || req.body.isolocal;
+            var isoweb = req.query.isoweb || req.body.isoweb;
+            var isoclient = req.query.isoclient || req.body.isoclient;
+
+            var source = isostore || isolocal || isoweb || isoclient;
+
+            assert.string(name, 'OS name');
+            assert.string(version, 'OS version');
+            assert.string(source, 'At least one iso source should be specified: ' +
+                'isostore, isolocal, isoweb');
+
+            var image = {};
+            image.id = uuid.v4();
+            image.name = name;
+            image.version = version;
+            image.status = 'OK';
+
+            return inventory.getOneImageByNameVersion(name, version)
+                .then(function(image){
+                    if(image) {
+                        throw new Errors.BadRequestError('Image already exists');
+                    }
+                })
+                .then(function() {
+                    if (isoclient) {
+                        image.iso = isoclient;
+                        return uploadIsoFile(req, res, image.iso);
+                    } else if(isolocal){
+                        image.iso = fsOp.getIsoNameFromStr(isolocal);
+                        return createSymbolLinkInStore(isolocal, image.iso);
+                    } else if(isoweb){
+                        image.iso = fsOp.getIsoNameFromWebLink(name, version, isoweb);
+                        return inventory.downloadIso(image, isoweb);
+                    } else if(isostore){
+                        image.iso = isostore;
+                        return Promise.resolve();
+                    }
+                })
```

Comment:
Is it there any possibility if other user request (maybe del, or another put?) happen during these two database operations? Will it create error behavior here?